git-sizer.go

```diff
@@ -306,22 +306,34 @@ func mainImplementation(ctx context.Context, stdout, stderr io.Writer, args []st
 		progressMeter = meter.NewProgressMeter(stderr, 100*time.Millisecond)
 	}
 
-	refRoots, err := sizes.CollectReferences(ctx, repo, rg)
-	if err != nil {
-		return fmt.Errorf("determining which reference to scan: %w", err)
-	}
-
-	roots := make([]sizes.Root, 0, len(refRoots)+len(flags.Args()))
-	for _, refRoot := range refRoots {
-		roots = append(roots, refRoot)
+	var roots []sizes.Root
+	var explicitRoots []sizes.Root
+	// If arguments are provided, use them as explicit roots.
+	if len(flags.Args()) > 0 {
+		explicitRoots = make([]sizes.Root, 0, len(flags.Args()))
+		for _, arg := range flags.Args() {
+			oid, err := repo.ResolveObject(arg)
+			if err != nil {
+				return fmt.Errorf("resolving command-line argument %q: %w", arg, err)
+			}
+			explicitRoots = append(explicitRoots, sizes.NewExplicitRoot(arg, oid))
+		}
 	}
 
-	for _, arg := range flags.Args() {
-		oid, err := repo.ResolveObject(arg)
+	// If no reference filters and no explicit roots were provided
+	if git.IsNoReferencesFilter(rgb.GetTopLevelGroup().GetFilter()) {
```

Comment:
Instead of revealing internal details via both `RefGroupBuilder.GetTopLevelGroup()` and `refGroup.GetFilter()` plus needing the strange function `IsNoReferencesFilter()`, how about adding a special-purpose method to `RefGroupBuilder`, something like the following:

```
// RefoptsSeen returns `true` iff any reference selection options were
// used. It is only valid after `Finish()` has been called.
func (rgb *RefGroupBuilder) RefoptsSeen() bool {
	return rgb.topLevelGroup.filter != git.NoReferencesFilter
}
```

`git.NoReferencesFilter` is a singleton, so I think that a simple comparison like this should suffice.